quarkchain/core.py

```diff
@@ -1162,49 +1171,76 @@ def __init__(self, tx_list):
+    def __init__(self, tx_list, version=1):
         self.tx_list = tx_list
-        self.version = version
+        self.version = 1  # force the version
 
     @staticmethod
-    def is_old_list(data):
-        bb = ByteBuffer(data)
-        size = bb.get_uint(4)
-        if size == 0:
-            return True
-        return (len(data) - 4) == CROSS_SHARD_TRANSACTION_DEPOSIT_DEPRECATED_SIZE * size
+    def get_version(data, size) -> int:
+        """-1 for deprecated list without version."""
+        if (
+            size == 0
+            or (len(data) - 4) == CROSS_SHARD_TRANSACTION_DEPOSIT_DEPRECATED_SIZE * size
+        ):
+            return -1
+        return int.from_bytes(data[-4:], byteorder="big")
 
-    @staticmethod
-    def from_data(data):
-        if not CrossShardTransactionList.is_old_list(data):
-            return CrossShardTransactionList.deserialize(data)
-
-        old_list = CrossShardTransactionDeprecatedList.deserialize(data)
-        return CrossShardTransactionList(
-            [
-                CrossShardTransactionDeposit(
-                    tx_hash=tx.tx_hash,
-                    from_address=tx.from_address,
-                    to_address=tx.to_address,
-                    value=tx.value,
-                    gas_price=tx.gas_price,
-                    gas_token_id=tx.gas_token_id,
-                    transfer_token_id=tx.transfer_token_id,
-                    gas_remained=0,
-                    message_data=b"",
-                    create_contract=False,
-                    is_from_root_chain=tx.is_from_root_chain,
-                    refund_rate=100,
-                )
-                for tx in old_list.tx_list
-            ]
-        )
+    @classmethod
+    def from_data(cls, data):
+        size = ByteBuffer(data).get_uint(4)
+        version = cls.get_version(data, size)
+
+        if version == 1:
+            return cls.deserialize(data)
+
+        if version == -1:
+            old_list = CrossShardTransactionDeprecatedList.deserialize(data)
+            return cls(
+                [
+                    CrossShardTransactionDeposit(
+                        tx_hash=tx.tx_hash,
+                        from_address=tx.from_address,
+                        to_address=tx.to_address,
+                        value=tx.value,
+                        gas_price=tx.gas_price,
+                        gas_token_id=tx.gas_token_id,
+                        transfer_token_id=tx.transfer_token_id,
+                        gas_remained=0,
+                        message_data=b"",
+                        create_contract=False,
+                        is_from_root_chain=tx.is_from_root_chain,
+                        refund_rate=100,
+                    )
+                    for tx in old_list.tx_list
+                ]
+            )
+
+        if version == 0:
+            v0_list = CrossShardTransactionListV0.deserialize(data)
+            # magic!
```

Comment:
I think we should use the similar logic for DeprecatedList for consistency.

Reply:
since we only added one field for V2, I think this is more readable and concise

Reply:
updated